### PR TITLE
libssh: #ifdef include files accordingly

### PR DESCRIPTION
**Pull request by bagder**
Reported-by: 梦终无痕
Bug: https://curl.se/mail/lib-2022-02/0131.html

### CODE REVIEW

**Comment by illusory-dream** [comment]
S_IFLNK and S_IFMT will be undefined variable without including unistd.h. Could it work as expecting if we define them manually.

**Comment by bagder** [comment]
> Could it work as expecting if we define them manually.

We can define them manually, sure. But isn't there a header file you can include that has them defined? Like maybe `windows.h` for example?

**Comment by bagder** [comment]
This fix seems to build on Windows now, at least with mingw

File: lib/vssh/libssh.c [hunk 1]
```diff
@@ -81,18 +81,33 @@
...
+#ifdef HAVE_FCNTL_H
 #include <fcntl.h>
+#endif
+
+#ifndef S_IFMT
+#define S_IFMT  00170000
+#endif
+#ifndef S_IFLNK
+#define S_IFLNK  0120000
+#endif
```
Comment (jay) [inline lib/vssh/libssh.c:105]: I don't think these should be defined, they may have different values on different platforms. For example my CRT uses 0xF000 for IFMT mask and there is no IFLNK file type. Also the code these are used for doesn't make sense to me:

~~~c
          if((sshc->readdir_attrs->flags & SSH_FILEXFER_ATTR_PERMISSIONS) &&
             ((sshc->readdir_attrs->permissions & S_IFMT) ==
              S_IFLNK)) {
~~~
So these are remote permissions? What if the client has IFMT, IFLNK values different from the server?

/cc @nmav

Comment (bagder) [reply lib/vssh/libssh.c:105]: Ah, I found these in liibssh/sftp.h:
~~~
#define SSH_S_IFMT   00170000
#define SSH_S_IFLNK  0120000

**Comment by gvanem** [comment]
But I noted that `WSAStartup()` is now called twice:
  * In libcurl's `Curl_win32_init()`.
  * And libssh's `ssh_socket_init()`.

Any way to avoid that?

**Comment by bagder** [comment]
It seems tricky because libssh has no apparent way to inhibit that. `ssh_socket_init()` is called from within `ssh_init()` but there's no way for us to tell it not to.

**Comment by jay** [comment]
> But I noted that `WSAStartup()` is now called twice:

that's fine as long as there's a corresponding cleanup call. multiple startup calls increment an internal counter like curl_global_init.

**Pull request closed without merging**